File: python/employee_scheduler.py
"""Weekly employee scheduler demonstrating conditionals, loops, and branching."""

from __future__ import annotations

import random
from dataclasses import dataclass, field

DAYS = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]
SHIFTS = ["morning", "afternoon", "evening"]
STAFF_PER_SHIFT = 2
MAX_DAYS = 5


@dataclass
class Employee:
    name: str
    # Each day stores ranked choices; index 0 is the first choice.
    preferences: dict[str, list[str]]
    days_worked: int = 0
    assignments: dict[str, str] = field(default_factory=dict)


def collect_employees() -> list[Employee]:
    """Collect a roster and optional ranked preferences from the keyboard."""
    while True:
        try:
            count = int(input("Number of employees (minimum 9): "))
            if count >= 9:
                break
        except ValueError:
            pass
        print("Please enter a whole number of at least 9.")

    employees: list[Employee] = []
    for number in range(1, count + 1):
        name = input(f"Employee {number} name: ").strip()
        while not name or any(employee.name.lower() == name.lower() for employee in employees):
            name = input("Enter a nonempty, unique name: ").strip()

        preferences: dict[str, list[str]] = {}
        print("Enter ranked shifts separated by commas (example: morning,evening,afternoon).")
        for day in DAYS:
            while True:
                choices = [item.strip().lower() for item in input(f"  {day}: ").split(",")]
                choices = [item for item in choices if item]
                if choices and len(choices) == len(set(choices)) and all(item in SHIFTS for item in choices):
                    preferences[day] = choices
                    break
                print("  Use one or more unique choices: morning, afternoon, evening.")
        employees.append(Employee(name, preferences))
    return employees


def sample_employees() -> list[Employee]:
    """Create repeatable demonstration data with ranked preferences."""
    names = ["Ava", "Ben", "Chloe", "Diego", "Emma", "Finn", "Grace", "Hassan", "Ivy"]
    rotations = [
        ["morning", "afternoon", "evening"],
        ["afternoon", "evening", "morning"],
        ["evening", "morning", "afternoon"],
    ]
    employees = []
    for employee_index, name in enumerate(names):
        preferences = {}
        for day_index, day in enumerate(DAYS):
            preferences[day] = rotations[(employee_index + day_index) % len(rotations)].copy()
        employees.append(Employee(name, preferences))
    return employees


def preference_rank(employee: Employee, day: str, shift: str) -> int:
    choices = employee.preferences.get(day, [])
    return choices.index(shift) if shift in choices else len(SHIFTS) + 1


def make_schedule(employees: list[Employee], seed: int = 42) -> dict[str, dict[str, list[str]]]:
    """Assign exactly two people per shift while respecting daily/weekly limits."""
    if len(employees) * MAX_DAYS < len(DAYS) * len(SHIFTS) * STAFF_PER_SHIFT:
        raise ValueError("At least 9 employees are needed to cover 42 shifts within the five-day limit.")

    rng = random.Random(seed)
    schedule = {day: {shift: [] for shift in SHIFTS} for day in DAYS}

    for day_index, day in enumerate(DAYS):
        # Lowest work count first prevents early employees from reaching five days too soon.
        eligible = [employee for employee in employees if employee.days_worked < MAX_DAYS]
        rng.shuffle(eligible)
        eligible.sort(key=lambda employee: employee.days_worked)

        for shift in SHIFTS:
            while len(schedule[day][shift]) < STAFF_PER_SHIFT:
                available = [employee for employee in eligible if day not in employee.assignments]
                if not available:
                    raise RuntimeError(f"Unable to cover {day} {shift}; add more employees.")

                # Prefer this shift, then favor the person with fewer total days.
                available.sort(
                    key=lambda employee: (preference_rank(employee, day, shift), employee.days_worked)
                )
                best_rank = preference_rank(available[0], day, shift)
                tied = [
                    employee
                    for employee in available
                    if preference_rank(employee, day, shift) == best_rank
                    and employee.days_worked == available[0].days_worked
                ]
                chosen = rng.choice(tied)
                schedule[day][shift].append(chosen.name)
                chosen.assignments[day] = shift
                chosen.days_worked += 1

        # This branch documents conflicts: employees not selected for a full preferred
        # shift remain eligible for another shift today, or naturally for the next day.
        assert all(len(schedule[day][shift]) == STAFF_PER_SHIFT for shift in SHIFTS)

    return schedule


def validate_schedule(employees: list[Employee], schedule: dict[str, dict[str, list[str]]]) -> None:
    """Raise AssertionError if any assignment rule is broken."""
    totals = {employee.name: 0 for employee in employees}
    for day in DAYS:
        assigned_today: list[str] = []
        for shift in SHIFTS:
            names = schedule[day][shift]
            assert len(names) >= STAFF_PER_SHIFT
            assigned_today.extend(names)
            for name in names:
                totals[name] += 1
        assert len(assigned_today) == len(set(assigned_today))
    assert all(total <= MAX_DAYS for total in totals.values())


def print_schedule(schedule: dict[str, dict[str, list[str]]], employees: list[Employee]) -> None:
    print("\nFINAL WEEKLY SCHEDULE")
    print("=" * 64)
    for day in DAYS:
        print(f"\n{day}")
        for shift in SHIFTS:
            print(f"  {shift.title():<10}: {', '.join(schedule[day][shift])}")
    print("\nDAYS WORKED")
    print("-" * 64)
    for employee in sorted(employees, key=lambda item: item.name):
        print(f"  {employee.name:<10}: {employee.days_worked}")


def main() -> None:
    choice = input("Use [S]ample data or [M]anual entry? ").strip().lower()
    employees = collect_employees() if choice.startswith("m") else sample_employees()
    schedule = make_schedule(employees)
    validate_schedule(employees, schedule)
    print_schedule(schedule, employees)


if __name__ == "__main__":
    main()

File: java/EmployeeScheduler.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.Scanner;

public class EmployeeScheduler {
    private static final List<String> DAYS = Arrays.asList(
            "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday");
    private static final List<String> SHIFTS = Arrays.asList("morning", "afternoon", "evening");
    private static final int STAFF_PER_SHIFT = 2;
    private static final int MAX_DAYS = 5;

    static class Employee {
        final String name;
        final Map<String, List<String>> preferences;
        final Map<String, String> assignments = new LinkedHashMap<>();
        int daysWorked = 0;

        Employee(String name, Map<String, List<String>> preferences) {
            this.name = name;
            this.preferences = preferences;
        }
    }

    static List<Employee> collectEmployees(Scanner scanner) {
        int count = 0;
        while (count < 9) {
            System.out.print("Number of employees (minimum 9): ");
            if (scanner.hasNextInt()) count = scanner.nextInt();
            else scanner.next();
            if (count < 9) System.out.println("Please enter a whole number of at least 9.");
        }
        scanner.nextLine();

        List<Employee> employees = new ArrayList<>();
        for (int number = 1; number <= count; number++) {
            String name;
            while (true) {
                System.out.print("Employee " + number + " name: ");
                name = scanner.nextLine().trim();
                final String proposedName = name;
                boolean duplicate = employees.stream().anyMatch(
                        employee -> employee.name.equalsIgnoreCase(proposedName));
                if (!name.isEmpty() && !duplicate) break;
                System.out.println("Enter a nonempty, unique name.");
            }

            Map<String, List<String>> preferences = new LinkedHashMap<>();
            System.out.println("Enter ranked shifts separated by commas.");
            for (String day : DAYS) {
                while (true) {
                    System.out.print("  " + day + ": ");
                    String[] pieces = scanner.nextLine().toLowerCase().split(",");
                    List<String> choices = new ArrayList<>();
                    for (String piece : pieces) {
                        String choice = piece.trim();
                        if (!choice.isEmpty()) choices.add(choice);
                    }
                    boolean valid = !choices.isEmpty()
                            && choices.stream().allMatch(SHIFTS::contains)
                            && choices.stream().distinct().count() == choices.size();
                    if (valid) {
                        preferences.put(day, choices);
                        break;
                    }
                    System.out.println("  Use unique choices: morning, afternoon, evening.");
                }
            }
            employees.add(new Employee(name, preferences));
        }
        return employees;
    }

    static List<Employee> sampleEmployees() {
        List<String> names = Arrays.asList("Ava", "Ben", "Chloe", "Diego", "Emma", "Finn", "Grace", "Hassan", "Ivy");
        List<List<String>> rotations = Arrays.asList(
                Arrays.asList("morning", "afternoon", "evening"),
                Arrays.asList("afternoon", "evening", "morning"),
                Arrays.asList("evening", "morning", "afternoon"));
        List<Employee> employees = new ArrayList<>();
        for (int employeeIndex = 0; employeeIndex < names.size(); employeeIndex++) {
            Map<String, List<String>> preferences = new LinkedHashMap<>();
            for (int dayIndex = 0; dayIndex < DAYS.size(); dayIndex++) {
                preferences.put(DAYS.get(dayIndex),
                        new ArrayList<>(rotations.get((employeeIndex + dayIndex) % rotations.size())));
            }
            employees.add(new Employee(names.get(employeeIndex), preferences));
        }
        return employees;
    }

    static int preferenceRank(Employee employee, String day, String shift) {
        List<String> choices = employee.preferences.getOrDefault(day, Collections.emptyList());
        int rank = choices.indexOf(shift);
        return rank >= 0 ? rank : SHIFTS.size() + 1;
    }

    static Map<String, Map<String, List<String>>> makeSchedule(List<Employee> employees, long seed) {
        int needed = DAYS.size() * SHIFTS.size() * STAFF_PER_SHIFT;
        if (employees.size() * MAX_DAYS < needed) {
            throw new IllegalArgumentException("At least 9 employees are needed to cover 42 shifts.");
        }

        Random random = new Random(seed);
        Map<String, Map<String, List<String>>> schedule = new LinkedHashMap<>();
        for (String day : DAYS) {
            Map<String, List<String>> daily = new LinkedHashMap<>();
            for (String shift : SHIFTS) daily.put(shift, new ArrayList<>());
            schedule.put(day, daily);
        }

        for (String day : DAYS) {
            List<Employee> eligible = new ArrayList<>();
            for (Employee employee : employees) {
                if (employee.daysWorked < MAX_DAYS) eligible.add(employee);
            }
            Collections.shuffle(eligible, random);
            eligible.sort(Comparator.comparingInt(employee -> employee.daysWorked));

            for (String shift : SHIFTS) {
                while (schedule.get(day).get(shift).size() < STAFF_PER_SHIFT) {
                    List<Employee> available = new ArrayList<>();
                    for (Employee employee : eligible) {
                        if (!employee.assignments.containsKey(day)) available.add(employee);
                    }
                    if (available.isEmpty()) {
                        throw new IllegalStateException("Unable to cover " + day + " " + shift);
                    }
                    available.sort(Comparator
                            .comparingInt((Employee employee) -> preferenceRank(employee, day, shift))
                            .thenComparingInt(employee -> employee.daysWorked));
                    int bestRank = preferenceRank(available.get(0), day, shift);
                    int bestDays = available.get(0).daysWorked;
                    List<Employee> tied = new ArrayList<>();
                    for (Employee employee : available) {
                        if (preferenceRank(employee, day, shift) == bestRank
                                && employee.daysWorked == bestDays) tied.add(employee);
                    }
                    Employee chosen = tied.get(random.nextInt(tied.size()));
                    schedule.get(day).get(shift).add(chosen.name);
                    chosen.assignments.put(day, shift);
                    chosen.daysWorked++;
                }
            }
        }
        return schedule;
    }

    static void validateSchedule(List<Employee> employees, Map<String, Map<String, List<String>>> schedule) {
        Map<String, Integer> totals = new LinkedHashMap<>();
        for (Employee employee : employees) totals.put(employee.name, 0);
        for (String day : DAYS) {
            List<String> assignedToday = new ArrayList<>();
            for (String shift : SHIFTS) {
                List<String> names = schedule.get(day).get(shift);
                if (names.size() < STAFF_PER_SHIFT) throw new AssertionError("Understaffed shift");
                assignedToday.addAll(names);
                for (String name : names) totals.put(name, totals.get(name) + 1);
            }
            if (assignedToday.stream().distinct().count() != assignedToday.size())
                throw new AssertionError("Employee assigned twice in one day");
        }
        for (int total : totals.values()) {
            if (total > MAX_DAYS) throw new AssertionError("Employee exceeds five days");
        }
    }

    static void printSchedule(Map<String, Map<String, List<String>>> schedule, List<Employee> employees) {
        System.out.println("\nFINAL WEEKLY SCHEDULE");
        System.out.println("================================================================");
        for (String day : DAYS) {
            System.out.println("\n" + day);
            for (String shift : SHIFTS) {
                String label = shift.substring(0, 1).toUpperCase() + shift.substring(1);
                System.out.printf("  %-10s: %s%n", label, String.join(", ", schedule.get(day).get(shift)));
            }
        }
        System.out.println("\nDAYS WORKED");
        System.out.println("----------------------------------------------------------------");
        employees.stream().sorted(Comparator.comparing(employee -> employee.name))
                .forEach(employee -> System.out.printf("  %-10s: %d%n", employee.name, employee.daysWorked));
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Use [S]ample data or [M]anual entry? ");
        String choice = scanner.nextLine().trim().toLowerCase();
        List<Employee> employees = choice.startsWith("m") ? collectEmployees(scanner) : sampleEmployees();
        Map<String, Map<String, List<String>>> schedule = makeSchedule(employees, 42L);
        validateSchedule(employees, schedule);
        printSchedule(schedule, employees);
        scanner.close();
    }
}
